File: src/spirite/base/util/glmath/Rect.java
package spirite.base.util.glmath;

public class Rect {
	public int x, y, width, height;

	public Rect( int width, int height) {
		x = y = 0;
		this.width = width;
		this.height = height;
	}
	public Rect( int x, int y, int width, int height) {
		this.x = x;
		this.y = y;
		this.width = width;
		this.height = height;
	}

	public Rect(Rect other) {
		this.x = other.x;
		this.y = other.y;
		this.width = other.width;
		this.height = other.height;
	}
	public Rect(Vec2i vec2i) {
		x = y = 0;
		this.width = vec2i.x;
		this.height = vec2i.y;
	}
	public boolean isEmpty() {
		return width <= 0 || height <= 0;
	}

	public Rect intersection(Rect other) {
		int x1 = Math.max( x,  other.x);
		int y1 = Math.max( y, other.y);
		int x2 = Math.min( x+width, other.x+other.width);
		int y2 = Math.min( y+height, other.y+other.height);
		return new Rect( x1, y1, x2-x1, y2-y1);
	}
	public boolean contains(int x2, int y2) {
		if( width <= 0 || height <= 0) return false;
		
		return !( x2 < x || y2 < y || x2 > x+width || y2 > y+height);
	}
	public boolean contains( Rect rect) {
		if( (width | height | rect.width | rect.height) < 0) return false;
		if( rect.x < this.x || rect.y < this.y) return false;
		int x2 = x + width;
		int X2 = rect.x + rect.width;
		if( X2 < rect.x ) {
			if( x2 >= x || X2 > x2) return false;
		} else {
			if( x2 >= x && X2 > x2) return false;
		}
		int y2 = y + height;
		int Y2 = rect.y + rect.height;
		if( Y2 < rect.y ) {
			if( y2 >= y || Y2 > y2) return false;
		} else {
			if( y2 >= y && Y2 > y2) return false;
		}
		return true;
	}
	public Rect union(Rect rect) {
		if( rect == null || rect.isEmpty()) return new Rect(this);
		if( isEmpty()) return new Rect(rect);
		
		return new Rect(
				Math.min( x, rect.x),
				Math.min( y, rect.y),
				Math.max( x + width, rect.x + rect.width),
				Math.max( y + height, rect.y + rect.height));
	}
	public Rect union(int x, int y, int w, int h) {
		return union(new Rect(x, y, w, h));
	}
	public boolean intersects(Rect r) {
         int tw = this.width;
         int th = this.height;
         int rw = r.width;
         int rh = r.height;
         if (rw <= 0 || rh <= 0 || tw <= 0 || th <= 0) {
             return false;
         }
         int tx = this.x;
         int ty = this.y;
         int rx = r.x;
         int ry = r.y;
         rw += rx;
         rh += ry;
         tw += tx;
         th += ty;
         //      overflow || intersect
         return ((rw < rx || rw > tx) &&
                 (rh < ry || rh > ty) &&
                 (tw < tx || tw > rx) &&
                 (th < ty || th > ry));
	}
}

File: src/main/kotlin/spirite/base/util/linear/Rect.kt
package spirite.base.util.linear

import kotlin.math.max
import kotlin.math.min

data class Rect(
        var x: Int,
        var y: Int,
        var width: Int,
        var height: Int
) {

    val isEmpty: Boolean get() = width <= 0 || height <= 0

    constructor(width: Int, height: Int): this(0,0,width,height) {}

    constructor(other: Rect) :this(other.x,other.y,other.width,other.height){}

    constructor(v: Vec2i) : this( 0, 0, v.x, v.y) {}

    infix fun intersection(other: Rect): Rect {
        val x1 = Math.max(x, other.x)
        val y1 = Math.max(y, other.y)
        val x2 = Math.min(x + width, other.x + other.width)
        val y2 = Math.min(y + height, other.y + other.height)
        return Rect(x1, y1, x2 - x1, y2 - y1)
    }

    fun contains(x2: Int, y2: Int): Boolean {
        return if (width <= 0 || height <= 0) false else !(x2 < x || y2 < y || x2 > x + width || y2 > y + height)
    }

    operator fun contains(rect: Rect): Boolean {
        if (width or height or rect.width or rect.height < 0) return false
        if (rect.x < this.x || rect.y < this.y) return false
        val x2 = x + width
        val X2 = rect.x + rect.width
        if (X2 < rect.x) {
            if (x2 >= x || X2 > x2) return false
        } else {
            if (x2 >= x && X2 > x2) return false
        }
        val y2 = y + height
        val Y2 = rect.y + rect.height
        if (Y2 < rect.y) {
            if (y2 >= y || Y2 > y2) return false
        } else {
            if (y2 >= y && Y2 > y2) return false
        }
        return true
    }

    infix fun union(rect: Rect?): Rect {
        return when {
            rect == null || rect.isEmpty -> Rect(this)
            isEmpty -> Rect(rect)
            else -> {
                val rx1 = min(x,rect.x)
                val ry1 = min(y,rect.y)
                val rx2 = max(x + width, rect.x + rect.width)
                val ry2 = max(y + height, rect.y + rect.height)
                Rect( rx1, ry1, rx2-rx1, ry2-ry1)
            }
        }
    }

    fun union(x: Int, y: Int, w: Int, h: Int): Rect {
        return union(Rect(x, y, w, h))
    }

    infix fun intersects(r: Rect): Boolean {
        var tw = this.width
        var th = this.height
        var rw = r.width
        var rh = r.height
        if (rw <= 0 || rh <= 0 || tw <= 0 || th <= 0) {
            return false
        }
        val tx = this.x
        val ty = this.y
        val rx = r.x
        val ry = r.y
        rw += rx
        rh += ry
        tw += tx
        th += ty
        //      overflow || intersect
        return (rw < rx || rw > tx) &&
                (rh < ry || rh > ty) &&
                (tw < tx || tw > rx) &&
                (th < ty || th > ry)
    }

    override fun toString() = "($x,$y),[$width x $height]"
}
